// ngx-malihu-scrollbar/dist/lib/malihu-scrollbar.directive.js
import { Directive, ElementRef, Input } from '@angular/core';
import { MalihuScrollbarService } from './malihu-scrollbar.service';
/**
 * Malihu Custom Scrollbar directive
 * Use this reference link for options definition
 * http://manos.malihu.gr/jquery-custom-content-scroller/
 */
var MalihuScrollbarDirective = (function () {
    function MalihuScrollbarDirective(elementRef, mScrollbarService) {
        this.elementRef = elementRef;
        this.mScrollbarService = mScrollbarService;
    }
    MalihuScrollbarDirective.prototype.ngAfterViewInit = function () {
        this.initElements();
        this.initScrollbar();
    };
    MalihuScrollbarDirective.prototype.ngOnDestroy = function () {
        this.destroyScrollbar();
    };
    MalihuScrollbarDirective.prototype.initElements = function () {
        this.scrollableElement = !!this.scrollElementId
            ? $("#" + this.scrollElementId)
            : $(this.elementRef.nativeElement);
        if (!!this.scrollElementId && this.scrollableElement.length === 0) {
            console.error("MalihuScrollbarDirective cannot find element with provided scrollElementId: " + this.scrollElementId + ".");
        }
    };
    MalihuScrollbarDirective.prototype.initScrollbar = function () {
        this.mScrollbarService.initScrollbar(this.scrollableElement, this.scrollbarOptions);
    };
    MalihuScrollbarDirective.prototype.destroyScrollbar = function () {
        try {
            this.mScrollbarService.destroy(this.scrollableElement);
        }
        catch (error) {
            // workaround for malihu-custom-scrollbar-plugin issue:
            // Cannot read property 'autoUpdate' of undefined
            // https://github.com/malihu/malihu-custom-scrollbar-plugin/issues/392
        }
    };
    return MalihuScrollbarDirective;
}());
export { MalihuScrollbarDirective };
MalihuScrollbarDirective.decorators = [
    { type: Directive, args: [{
                selector: '[MalihuScrollbar], [malihu-scrollbar]',
            },] },
];
/** @nocollapse */
MalihuScrollbarDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: MalihuScrollbarService, },
]; };
MalihuScrollbarDirective.propDecorators = {
    'scrollElementId': [{ type: Input },],
    'scrollbarOptions': [{ type: Input },],
};
