// ngx-malihu-scrollbar/dist/bundles/ngx-malihu-scrollbar.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core'], factory) :
	(factory((global.ng = global.ng || {}, global.ng.malihuscrollbar = {}),global.ng.core));
}(this, (function (exports,core) { 'use strict';

var MalihuScrollbarService = (function () {
    function MalihuScrollbarService(zone) {
        this.zone = zone;
    }
    MalihuScrollbarService.prototype.initScrollbar = function (element, options) {
        var jQueryElement = this.getElement(element);
        this.zone.runOutsideAngular(function () { return jQueryElement.mCustomScrollbar(options); });
        if (jQueryElement.length > 0 && jQueryElement[0].tagName === 'BODY') {
            jQueryElement[0].style.position = 'absolute';
            jQueryElement[0].style.overflow = 'auto';
            jQueryElement[0].style.height = '100vh';
            jQueryElement[0].style.width = '100vw';
        }
    };
    MalihuScrollbarService.prototype.scrollTo = function (element, parameter, options) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('scrollTo', parameter, options); });
    };
    MalihuScrollbarService.prototype.update = function (element) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('update'); });
    };
    MalihuScrollbarService.prototype.stop = function (element) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('stop'); });
    };
    MalihuScrollbarService.prototype.disable = function (element, reset) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('disable', !!reset); });
    };
    MalihuScrollbarService.prototype.destroy = function (element) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('destroy'); });
    };
    MalihuScrollbarService.prototype.getElement = function (element) {
        if (typeof element === 'string' || element instanceof String) {
            return $(element);
        }
        if (typeof element === 'object' && element instanceof HTMLElement) {
            return $(element);
        }
        if (element instanceof jQuery || 'jquery' in Object(element)) {
            return element;
        }
        throw Error("Unsupported element type in MalihuScrollbarService: " + element);
    };
    return MalihuScrollbarService;
}());
MalihuScrollbarService.decorators = [
    { type: core.Injectable },
];
/** @nocollapse */
MalihuScrollbarService.ctorParameters = function () { return [
    { type: core.NgZone, },
]; };

/**
 * Malihu Custom Scrollbar directive
 * Use this reference link for options definition
 * http://manos.malihu.gr/jquery-custom-content-scroller/
 */
var MalihuScrollbarDirective = (function () {
    function MalihuScrollbarDirective(elementRef, mScrollbarService) {
        this.elementRef = elementRef;
        this.mScrollbarService = mScrollbarService;
    }
    MalihuScrollbarDirective.prototype.ngAfterViewInit = function () {
        this.initElements();
        this.initScrollbar();
    };
    MalihuScrollbarDirective.prototype.ngOnDestroy = function () {
        this.destroyScrollbar();
    };
    MalihuScrollbarDirective.prototype.initElements = function () {
        this.scrollableElement = !!this.scrollElementId
            ? $("#" + this.scrollElementId)
            : $(this.elementRef.nativeElement);
        if (!!this.scrollElementId && this.scrollableElement.length === 0) {
            console.error("MalihuScrollbarDirective cannot find element with provided scrollElementId: " + this.scrollElementId + ".");
        }
    };
    MalihuScrollbarDirective.prototype.initScrollbar = function () {
        this.mScrollbarService.initScrollbar(this.scrollableElement, this.scrollbarOptions);
    };
    MalihuScrollbarDirective.prototype.destroyScrollbar = function () {
        try {
            this.mScrollbarService.destroy(this.scrollableElement);
        }
        catch (error) {
            // workaround for malihu-custom-scrollbar-plugin issue:
            // Cannot read property 'autoUpdate' of undefined
            // https://github.com/malihu/malihu-custom-scrollbar-plugin/issues/392
        }
    };
    return MalihuScrollbarDirective;
}());
MalihuScrollbarDirective.decorators = [
    { type: core.Directive, args: [{
                selector: '[MalihuScrollbar], [malihu-scrollbar]',
            },] },
];
/** @nocollapse */
MalihuScrollbarDirective.ctorParameters = function () { return [
    { type: core.ElementRef, },
    { type: MalihuScrollbarService, },
]; };
MalihuScrollbarDirective.propDecorators = {
    'scrollElementId': [{ type: core.Input },],
    'scrollbarOptions': [{ type: core.Input },],
};

var MalihuScrollbarModule = (function () {
    function MalihuScrollbarModule() {
    }
    MalihuScrollbarModule.forRoot = function () {
        return {
            ngModule: MalihuScrollbarModule,
            providers: [MalihuScrollbarService],
        };
    };
    MalihuScrollbarModule.forChild = function () {
        return {
            ngModule: MalihuScrollbarModule,
        };
    };
    return MalihuScrollbarModule;
}());
MalihuScrollbarModule.decorators = [
    { type: core.NgModule, args: [{
                exports: [MalihuScrollbarDirective],
                declarations: [MalihuScrollbarDirective],
            },] },
];
/** @nocollapse */
MalihuScrollbarModule.ctorParameters = function () { return []; };

exports.MalihuScrollbarService = MalihuScrollbarService;
exports.MalihuScrollbarDirective = MalihuScrollbarDirective;
exports.MalihuScrollbarModule = MalihuScrollbarModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
